pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -1665,73 +1665,91 @@ private void parseQuery() {
+
+        default:
+          rowBuffer.set(columnIndex, connection.encodeString(String.valueOf(valueObject)));
+          break;
+      }
+
+    }
+  }
+
+  private void updateRowBuffer(boolean isInsert) throws SQLException {
+
     for (Map.Entry<String, Object> entry : updateValues.entrySet()) {
       int columnIndex = findColumn(entry.getKey()) - 1;
-
       Object valueObject = entry.getValue();
-      if (valueObject instanceof PGobject) {
-        String value = ((PGobject) valueObject).getValue();
-        rowBuffer.set(columnIndex, (value == null) ? null : connection.encodeString(value));
-      } else {
-        switch (getSQLType(columnIndex + 1)) {
-
-          // boolean needs to be formatted as t or f instead of true or false
-          case Types.BIT:
-          case Types.BOOLEAN:
-            rowBuffer.set(columnIndex, connection
-                .encodeString(((Boolean) valueObject).booleanValue() ? "t" : "f"));
-            break;
-            //
-            // toString() isn't enough for date and time types; we must format it correctly
-            // or we won't be able to re-parse it.
-            //
-          case Types.DATE:
-            rowBuffer.set(columnIndex, connection
-                .encodeString(
-                    connection.getTimestampUtils().toString(
-                        getDefaultCalendar(), (Date) valueObject)));
-            break;
-
-          case Types.TIME:
-            rowBuffer.set(columnIndex, connection
-                .encodeString(
-                    connection.getTimestampUtils().toString(
-                        getDefaultCalendar(), (Time) valueObject)));
-            break;
-
-          case Types.TIMESTAMP:
-            rowBuffer.set(columnIndex, connection.encodeString(
-                connection.getTimestampUtils().toString(
-                    getDefaultCalendar(), (Timestamp) valueObject)));
-            break;
-
-          case Types.NULL:
-            // Should never happen?
-            break;
-
-          case Types.BINARY:
-          case Types.LONGVARBINARY:
-          case Types.VARBINARY:
-            if (isBinary(columnIndex + 1)) {
-              rowBuffer.set(columnIndex, (byte[]) valueObject);
-            } else {
-              try {
-                rowBuffer.set(columnIndex,
-                    PGbytea.toPGString((byte[]) valueObject).getBytes("ISO-8859-1"));
-              } catch (UnsupportedEncodingException e) {
-                throw new PSQLException(
-                    GT.tr("The JVM claims not to support the encoding: {0}", "ISO-8859-1"),
-                    PSQLState.UNEXPECTED_ERROR, e);
-              }
-            }
-            break;
+      setRowBufferColumn(columnIndex, valueObject);
+    }
 
-          default:
-            rowBuffer.set(columnIndex, connection.encodeString(String.valueOf(valueObject)));
-            break;
-        }
+    if (isInsert) {
+      final ResultSet generatedKeys = insertStatement.getGeneratedKeys();
```

Comment:
Should it be closed via `#close()` ?

Reply:
Good catch, I suppose it should be closed. Fix pushed.

Reply:
How about try-finally?

Reply:
We can be more defensive, I was trying to match the style of the surrounding code, but perhaps I missed the target. Change pushed.